test/functional/core/exit_spec.lua

```diff
@@ -43,4 +43,29 @@ describe('v:exiting', function()
     run(on_request, nil, on_setup)
   end)
 
+  it('is specified a non-zero exit code after :cquit', function()
+    local function on_setup()
+      command('autocmd VimLeavePre * call rpcrequest('..cid..', "")')
+      command('autocmd VimLeave    * call rpcrequest('..cid..', "")')
+      command('cquit 123')
+    end
+    local function on_request()
+      eq(123, eval('v:exiting'))
+      return ''
+    end
+    run(on_request, nil, on_setup)
+  end)
+
+  it('is specified a zero exit code after :cquit', function()
+    local function on_setup()
+      command('autocmd VimLeavePre * call rpcrequest('..cid..', "")')
+      command('autocmd VimLeave    * call rpcrequest('..cid..', "")')
+      command('cquit 0')
+    end
+    local function on_request()
+      eq(0, eval('v:exiting'))
+      return ''
+    end
+    run(on_request, nil, on_setup)
+  end)
```

Comment:
Still missing most of the tests I listed. I would also highly advise on creating a “test generator” function called like `test_cq('cquit 0', 0, nil)`:

```lua
local function test_cq(cmdline, exit_code, redir_msg)
  if redir_msg then
    eq('\n' .. redir_msg, redir_exec(cmdline))
    wait()
    eq(1, eval('1'))  -- Check that Neovim did not crash or exit
  else
    funcs.system({nvim_prog, '-u', 'NONE', '-i', 'NONE', '--headless', '--cmd', cmdline})
    eq(exit_code, eval('v:shell_error'))
    local function on_setup()
       command('autocmd VimLeavePre * call rpcrequest('..cid..', "")')
       command('autocmd VimLeave    * call rpcrequest('..cid..', "")')
       command(cmdline)
     end
     local function on_request()
       eq(exit_code, eval('v:exiting'))
       return ''
     end
     run(on_request, nil, on_setup)
  end
end
```

Note the significant difference: I do not suggest checking *only* `v:exiting`, this test does not really serve its purpose because you can’t observe exit code before Neovim actually exits and you can’t have `v:exiting` available after it exits: test is only valid as long as there is connection between `v:exiting` and exit value and it is not necessary the case if somebody added a bug. Actually since I mentioned it: what you are testing is `:cquit`, *not* `v:exiting`, so while checking for it sideways is fine, setting `v:exiting` is by no means a purpose of `:cquit`: it is purpose of `v:exiting` to contain exit value Neovim will use, purpose of `:cq` is quit Neovim with specified exit value which is checked by `system()` check only.

And remove `:cq` out of the current `describe()` block, it is not correct.